fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,226 @@
+        logits /= self.temperature
+        labels = torch.zeros(features.size(0)).to(self.device).long()
+
+        return self.ce_criterion(logits, labels)
+
+    def get_contrastive_loss(
+        self, local_features: torch.Tensor, old_local_features: torch.Tensor, global_features: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Compute contrastive over current local features (z_p) with old local features (hat{z_p}) as positive pairs
+        and current global features (z_s) as negative pairs.
+        """
+        return self.compute_contrastive_loss(
+            features=local_features,  # (z_p)
+            positive_pairs=old_local_features,  # (\hat{z_p})
+            negative_pairs=global_features,  # (z_s)
+        )
+
+    def get_perfcl_loss(
+        self,
+        local_features: torch.Tensor,
+        old_local_features: torch.Tensor,
+        global_features: torch.Tensor,
+        old_global_features: torch.Tensor,
+        aggregated_global_features: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Perfcl loss consists of two contrastive loss computation.
+
+        First one aims to enhance the similarity between the current global features (z_s) and aggregated global
+        features (z_g) as positive pairs while reducing the similarity between the current global features (z_s)
+        and old global features (hat{z_s}) as negative pairs.
+
+        Second one aims to enhance the similarity between the current local features (z_p) and old local features
+        (hat{z_p}) as positive pairs while reducing the similarity between the current local features (z_p) and
+        aggregated lobal features (z_g) as negative pairs.
+        """
+
+        contrastive_loss_minimize = self.compute_contrastive_loss(
+            features=global_features,  # (z_s)
+            positive_pairs=aggregated_global_features,  # (z_g)
+            negative_pairs=old_global_features,  # (\hat{z_s})
+        )
+        contrastive_loss_maximize = self.compute_contrastive_loss(
+            features=local_features,  # (z_p)
+            positive_pairs=old_local_features,  # (\hat{z_p})
+            negative_pairs=aggregated_global_features,  # (z_g)
+        )
+
+        return contrastive_loss_minimize, contrastive_loss_maximize
+
+    def compute_loss(self, preds: Dict[str, torch.Tensor], target: torch.Tensor) -> Losses:
+
+        loss = self.criterion(preds["prediction"], target)
+        total_loss = loss
+        additional_losses = {}
+
+        # Optimal cos_sim_loss_weight for FedIsic dataset is 100.0
+        if self.cos_sim_loss_weight:
+            cos_sim_loss = self.get_cosine_similarity_loss(
+                local_features=preds["local_features"],
+                global_features=preds["global_features"],
+            )
+            total_loss += self.cos_sim_loss_weight * cos_sim_loss
+            additional_losses["cos_sime_loss"] = cos_sim_loss
```

Comment:
I'm not 100% sure, but I imagine that the key `"cos_sime_loss"` might cause some issues? Either way I think the typo should probably be fixed.